Retune fixes

glenvt18 opened the pull request:
This PR fixes two issues with re-tune:

1.  When a higher priority recording or streaming task is about to zap to another transponder on the device from which VNSI is currently streaming, it detaches all receivers from the device, which, in turn, triggers re-tune. Since VDR has no synchronization mechanism for zapping (`GetDevice()->SwitchChannel()->AttachReceiver()` chain), a race occurs. The result is usually that the recording calls an emergency exit after not receiving TS packets for 30 seconds. To reproduce this, start streaming a channel, then start recording (or another VNSI streaming with a higher priority) of a channel from another transponder. VDR should restart in 30 seconds. The proposed solution is to disable re-tune in such cases. Even if (when) VDR had zapping synchronization, re-tuning after the device has been switched to another transponder by a higher priority task wouldn't be practical anyway.
To detect whether to re-tune or not, a helper dummy receiver is attached to the device. If this dummy receiver is detached, re-tuning should be disabled.

2.  `DisableScrambleTimeout` option only works for clients with priority < 0 (after 9a8fef72 it's not possible anymore). For CAMs like dvbapi, re-tuning after scrambling timeout has expired doesn't help because the channel is then being marked 'checked' (blacklisted) for 15 seconds. To reproduce this, add a delay > 3 seconds to the CAM, or switch it off, and, after some time, switch it on. Solution: 'uncheck' before zapping (`GetDevice()` call).

FernetMenta commented:
thanks for contributing. some questions:

> DisableScrambleTimeout option only works for clients with priority < 0

where do you see this relationship? why does it not work for prio 0 and higher?

glenvt18 commented:
It's in device.c:

```C++
for (int i = 0; i < MAXRECEIVERS; i++) {
    if (receiver[i] && receiver[i]->WantsPid(Pid)) {
       if (DetachReceivers && cs && (!cs->IsActivating() || receiver[i]->Priority() >= LIVEPRIORITY)) {
          dsyslog("detaching receiver - won't decrypt channel %s with CAM %d", *receiver[i]->ChannelID().ToString(), CamSlotNumber);
          ChannelCamRelations.SetChecked(receiver[i]->ChannelID(), CamSlotNumber);
          Detach(receiver[i]);
          }
       else
          receiver[i]->Receive(b, TS_SIZE);
       if (DescramblingOk)
          ChannelCamRelations.SetDecrypt(receiver[i]->ChannelID(), CamSlotNumber);
       }
    }
```
`LIVEPRIORITY` is 0. You've added `cs->StartActivation()`, so `cs->IsActivating()` is true. But, if `receiver[i]->Priority() >= LIVEPRIORITY`, Detach() is inevitable. Notice `SetChecked()` call above `Detach()`. That's 'blacklisting'. `ClrChecked()` is the opposite. In practice, after a channel has been blacklisted (scrambling timeout expired), it is not possible to decrypt at all. But with `ClrChecked()` a series of re-tunes (usually 2 are enough) allows to zap successfully.
A better solution would be to have a virtual function `int cCamSlot::ScramblingTimeout()` instead of using hard-coded 3 sec value. In this case, CAM implementations, which don't need slot enumeration, could return 0.
But, for now, all we can do is this workaround.

BTW. I've installed kodi-pvr-vdr-vnsi 2.6.5-1 which was built 2016-09-18 from ppa:team-xbmc/unstable. This addon uses protocol 9 and always transmits priority=-1 regardless of the KODI setting. Is it a bug? So, when I set priority=5, it is transmitted as -1, and then, finally, converted to 0. That's why I added priority logging :-)

FernetMenta commented:
Right, that's the reason I handled deactivation of scramble timeout differently in my first version but somehow seem to forgot it :)

FernetMenta commented:
I will investigate the issue with priority.

FernetMenta commented on videoinput.c:
```diff
@@ -564,9 +635,23 @@ inline void cVideoInput::Receive(const uchar *data, int length)
 }
 
-void cVideoInput::Retune()
+void cVideoInput::RequestRetune()
```
this can be called from 2 different threads. either we need a lock or m_RetuneRequested should be atomic

glenvt18 replied:
Strictly, yes.

FernetMenta commented on videoinput.c:
```diff
@@ -564,9 +635,23 @@ inline void cVideoInput::Receive(const uchar *data, int length)
+    (void)m_Device->Receiving();  // wait for the receivers mutex
+    if (!m_DummyReceiver->IsAttached())  // DetachAllReceivers() was called
+      return CLOSE;
+    else if (!m_PmtChange)
```
why do you want to wait here?
m_PmtChange needs a lock to should be atomic

glenvt18 replied:
Receivers operations (`AttachReceiver()`, `DetachXXX()`, `Receiving()`) are protected by a mutex, but `IsAttached()` is not. It just reads `device` value. At the moment we get `m_Receiver->device == NULL` (`!IsAttached()`), `m_DummyReceiver->device` might be not NULL - `DetachAllReceivers()` hasn't yet detached it. That's why we have to wait for `DetachAllReceivers()` to complete. At this point we can be sure that if both receivers has been detached, we get two NULL `device`s. In other words, without `Receiving()` (or any other call that holds the receiver mutex) it's possible to get a false RETUNE when it should have been CLOSE. This would clash with recording.
But `device` members are not atomic/protected as well as `m_RetuneRequested`. We can use our own 'protected' flags set by receiver callbacks instead. I'll rework this and set a fix as a new PR. Do you agree?
Do we still need m_PmtChange?

FernetMenta commented on videoinput.c:
```diff
@@ -492,6 +558,9 @@ void cVideoInput::Close()
       cCamSlot *cs = m_Device->CamSlot();
       if (cs)
         cs->CancelActivation();
+      if (m_Receiver)
```
I have created a new setting for this strategy. Unfortunately it does not work reliable because when the device blacklists the channel it also unassigns the cam. As a result cs is null. So it only works for camSlot 0.
We need to store the camslot number on open. Working on it and push in an hour or so for you to review.

glenvt18 replied:
I can't test it with more than one slot. That VDR code looks really nasty.

And I'm working on re-tune...

FernetMenta replied:
yes, VDR gives us a hard time on this.
 PR submitted.
